Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp

```diff
@@ -883,16 +899,73 @@ GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin,
-            GUniqueOutPtr<GError> error;
-            result = gst_parse_bin_from_description("aacparse ! capsfilter caps=\"audio/mpeg, stream-format=(string)adts\"", TRUE, &error.outPtr());
-            if (result)
-                gst_object_set_name(GST_OBJECT_CAST(result.get()), parserName.ascii().data());
-            else
-                GST_WARNING_OBJECT(bin, "Unable to create AAC parser: %s", error->message);
+            // Replace explicit signalling of SBR (HE AAC) and PS (HE AACv2) with implicit
+            // signalling (i.e. let the decoder figure out what AAC extensions are used).
+            // Unfortunately it is not uncommon for AAC streams to explicitly signal extensions they
+            // don't actually use, which is known to cause problems in avdec_aac.
+
+            result = makeGStreamerElement("aacparse"_s, parserName);
+            if (!result)
+                break;
+
+            gst_pad_add_probe(demuxerSrcPad.get(), GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM, reinterpret_cast<GstPadProbeCallback>(+[]([[maybe_unused]] GstPad* pad, GstPadProbeInfo* info, gpointer) -> GstPadProbeReturn {
+                auto event = GST_PAD_PROBE_INFO_EVENT(info);
+                if (GST_EVENT_TYPE(event) != GST_EVENT_CAPS)
+                    return GST_PAD_PROBE_OK;
+
+                GstCaps* caps;
+                gst_event_parse_caps(event, &caps);
+
+                GstStructure* structure = gst_caps_get_structure(caps, 0);
+                auto mediaType = gstStructureGetName(structure);
+                ASSERT(mediaType == "audio/mpeg"_s);
+                if (mediaType != "audio/mpeg"_s) [[unlikely]]
+                    return GST_PAD_PROBE_REMOVE;
+
+                int mpegVersion = gstStructureGet<int>(structure, "mpegversion"_s).value_or(0);
+                ASSERT(mpegVersion == 4);
+                if (mpegVersion != 4) [[unlikely]]
+                    return GST_PAD_PROBE_REMOVE;
+
+                if (!gst_structure_has_field_typed(structure, "codec_data", GST_TYPE_BUFFER)) [[unlikely]]
+                    return GST_PAD_PROBE_REMOVE;
+
+                GRefPtr<GstBuffer> codecDataBuffer;
+                gst_structure_get(structure, "codec_data", GST_TYPE_BUFFER, &codecDataBuffer.outPtr(), nullptr);
+                GstMappedBuffer data(codecDataBuffer, GST_MAP_READ);
+                GstBitReader reader;
+                gst_bit_reader_init(&reader, data.data(), static_cast<unsigned>(data.size()));
+                auto audioObjectType = aacGetAudioObjectType(reader);
+
+                if (!audioObjectType || ((*audioObjectType != 5) && (*audioObjectType != 29)))
+                    return GST_PAD_PROBE_OK;
+
+                GST_DEBUG_OBJECT(pad, "AAC SBR or PS object type detected in codec data, replacing it with a LC object type. The decoder should be able to handle this properly");
+
+                auto channels = gst_codec_utils_aac_get_channels(data.data(), data.size());
```

Comment:
I was concerned that gst_codec_utils_aac_get_channels() could be too high level and actually parse PS as well, but it doesn't. So as long as it is not changed afterwards, this would work.

Personally I would have just read the bits myself since it's not more difficult than reading the frequency bits.

Reply:
Well I think we can consider that pbutils API behaviour stable, its doc clearly implies the value return should conform with the spec.